File: android/app/src/main/java/com/vaultsync/domain/usecase/ImportDocumentUseCase.kt
package com.vaultsync.domain.usecase

import com.vaultsync.domain.model.AppError
import com.vaultsync.domain.model.Document
import com.vaultsync.domain.model.OperationType
import com.vaultsync.domain.model.SyncOperation
import com.vaultsync.domain.model.SyncStatus
import com.vaultsync.domain.repository.CryptoService
import com.vaultsync.domain.repository.DocumentRepository
import com.vaultsync.domain.repository.FileStorage
import com.vaultsync.domain.repository.SyncRepository
import java.io.ByteArrayOutputStream
import java.util.UUID

class ImportDocumentUseCase(
    private val documentRepository: DocumentRepository,
    private val syncRepository: SyncRepository,
    private val fileStorage: FileStorage,
    private val cryptoService: CryptoService
) {
    suspend operator fun invoke(
        name: String,
        mimeType: String,
        bytes: ByteArray
    ): Document {
        if (bytes.isEmpty()) {
            throw AppError.StorageError("Cannot import empty file: $name")
        }

        // 1. Calculate SHA-256 checksum
        val sha256 = cryptoService.calculateSha256(bytes)

        // 2. Check for duplicate by SHA-256
        val existing = documentRepository.findBySha256(sha256)
        if (existing != null) {
            throw AppError.DuplicateError("A document with identical content already exists: ${existing.name}")
        }

        // 3. Encrypt payload with hardware key
        val encryptedPayload = cryptoService.encrypt(bytes)

        // Package IV + ciphertext for atomic storage
        val combinedOutput = ByteArrayOutputStream()
        combinedOutput.write(encryptedPayload.iv.size)
        combinedOutput.write(encryptedPayload.iv)
        combinedOutput.write(encryptedPayload.ciphertext)
        val encryptedBytes = combinedOutput.toByteArray()

        val docId = UUID.randomUUID().toString()
        val storagePath = "vault_$docId.enc"

        // 4. Write encrypted payload to sandboxed storage
        fileStorage.writeFile(storagePath, encryptedBytes)

        val now = System.currentTimeMillis()
        val document = Document(
            id = docId,
            name = name,
            size = bytes.size.toLong(),
            mimeType = mimeType,
            localPath = storagePath,
            sha256 = sha256,
            createdAt = now,
            modifiedAt = now,
            syncStatus = SyncStatus.PENDING
        )

        // 5. Persist document metadata
        documentRepository.saveDocument(document)

        // 6. Record sync operation in durable queue
        val syncOp = SyncOperation(
            id = UUID.randomUUID().toString(),
            documentId = docId,
            operationType = OperationType.UPLOAD,
            status = com.vaultsync.domain.model.OperationStatus.PENDING,
            startedAt = now
        )
        syncRepository.recordOperation(syncOp)

        return document
    }
}

File: ios/Sources/VaultSync/Domain/UseCases/ImportDocumentUseCase.swift
import Foundation

public final class ImportDocumentUseCase: Sendable {
    private let documentRepository: any DocumentRepository
    private let syncRepository: any SyncRepository
    private let fileStorage: any FileStorage
    private let cryptoService: any CryptoService

    public init(
        documentRepository: any DocumentRepository,
        syncRepository: any SyncRepository,
        fileStorage: any FileStorage,
        cryptoService: any CryptoService
    ) {
        self.documentRepository = documentRepository
        self.syncRepository = syncRepository
        self.fileStorage = fileStorage
        self.cryptoService = cryptoService
    }

    public func execute(name: String, mimeType: String, bytes: Data) async throws -> Document {
        guard !bytes.isEmpty else {
            throw AppError.storageError("Cannot import empty file: \(name)")
        }

        let sha256 = await cryptoService.calculateSha256(data: bytes)

        if let existing = try await documentRepository.findBySha256(sha256) {
            throw AppError.duplicateError("A document with identical content already exists: \(existing.name)")
        }

        let payload = try await cryptoService.encrypt(data: bytes, keyAlias: cryptoService.defaultKeyAlias)

        var combined = Data()
        var ivLength = UInt8(payload.iv.count)
        combined.append(&ivLength, count: 1)
        combined.append(payload.iv)
        combined.append(payload.ciphertext)

        let docId = UUID().uuidString
        let storagePath = "vault_\(docId).enc"

        _ = try await fileStorage.writeFile(at: storagePath, data: combined)

        let now = Date()
        let document = Document(
            id: docId,
            name: name,
            size: Int64(bytes.count),
            mimeType: mimeType,
            localPath: storagePath,
            sha256: sha256,
            createdAt: now,
            modifiedAt: now,
            syncStatus: .pending
        )

        try await documentRepository.saveDocument(document)

        let syncOp = SyncOperation(
            id: UUID().uuidString,
            documentId: docId,
            operationType: .upload,
            status: .pending,
            startedAt: now
        )
        try await syncRepository.recordOperation(syncOp)

        return document
    }
}
